src/write/file.rs

```diff
@@ -195,15 +219,23 @@ impl<W: Write> FileWriter<W> {
             None,
         );
 
-        let len = end_file(&mut self.writer, metadata)?;
+        let len = end_file(&mut self.writer, &metadata)?;
         self.state = State::Finished;
+        self.metadata = Some(metadata);
         Ok(self.offset + len)
     }
 
     /// Returns the underlying writer.
     pub fn into_inner(self) -> W {
         self.writer
     }
+
+    /// Returns the underlying writer and [`FileMetaData`]
+    /// # Panics
+    /// This function panics if [`Self::end`] has not yet been called
+    pub fn into_inner_and_metadata(self) -> (W, FileMetaData) {
```

Comment:
One thing I didn't realize originally (because they're both named `FileMetaData`) is that this returns the _thrift_ `FileMetaData` and not the `parquet2` `FileMetaData`.

I have a slight feeling that returning the `parquet2` `FileMetaData` would be cleaner here, so that applications don't have to mix and match the two structs. Thoughts?

Reply:
I agree. I tried that, but it is not easy - `FileMetadata` is a "read-specialized" struct - it contains column descriptors derived (transversed from the parquet schema tree) from the parquet schema that are used across all row groups.

To expose `FileMetadata` when writing requires doing the same for the sole purpose of this feature, which I considered wasteful.

In other words, I see the need for 3 structs:

* "ThriftFileMetadata" - what we convert from and to bytes
* "ReadFileMetadata" - "annotated row groups from transversing the schema in ThriftFileMetadata" + "validated ThriftFileMetadata (e.g. num_rows is `usize`)"
* "WrittenFileMetadata" - "validated ThriftFileMetadata (e.g. num_rows is `usize`)"

from which "ThriftFileMetadata" should be an implementation detail of this crate. But I do not have good idea how to do this. Ideas welcome :)

Reply:
Note that you can perform the conversion using `FileMetadata::try_from_thrift`, which performs the conversion and a some validation